Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrElementBuilders.kt

```diff
@@ -0,0 +1,912 @@
+                functions = emptyList(),
+                modality = ClassStubModality.NONE,
+                annotations = listOfNotNull(structAnnotation),
+                superClassInit = superClassInit,
+                constructorParams = listOf(rawPtrConstructorParam),
+                companion = companion
+        ))
+    }
+
+    private fun getArrayLength(type: ArrayType): Long {
+        val unwrappedElementType = type.elemType.unwrapTypedefs()
+        val elementLength = if (unwrappedElementType is ArrayType) {
+            getArrayLength(unwrappedElementType)
+        } else {
+            1L
+        }
+
+        val elementCount = when (type) {
+            is ConstArrayType -> type.length
+            is IncompleteArrayType -> 0L
+            else -> TODO(type.toString())
+        }
+
+        return elementLength * elementCount
+    }
+
+    private tailrec fun Type.isIntegerTypeSigned(): Boolean = when (this) {
+        is IntegerType -> this.isSigned
+        is BoolType -> false
+        is EnumType -> this.def.baseType.isIntegerTypeSigned()
+        is Typedef -> this.def.aliased.isIntegerTypeSigned()
+        else -> error(this)
+    }
+
+    /**
+     * Produces to [out] the definition of Kotlin class representing the reference to given forward (incomplete) struct.
+     */
+    private fun generateForwardStruct(s: StructDecl): List<StubIrElement> = when (context.platform) {
+        KotlinPlatform.JVM -> {
+            // ("class ${s.kotlinName.asSimpleName()}(rawPtr: NativePtr) : COpaque(rawPtr)")
+            val classifier = context.getKotlinClassForPointed(s)
+            val superClass = SymbolicStubType("COpaque")
+            val rawPtrConstructorParam = ConstructorParamStub("rawPtr", SymbolicStubType("NativePtr"))
+            val superClassInit = SuperClassInit(superClass, listOf(rawPtrConstructorParam))
+            val origin = StubOrigin.Struct(s)
+            listOf(ClassStub.Simple(classifier, ClassStubModality.NONE, listOf(rawPtrConstructorParam), superClassInit, origin = origin))
+        }
+        KotlinPlatform.NATIVE -> emptyList()
+    }
+}
+
+internal class EnumStubBuilder(
+        override val context: StubsBuildingContext,
+        private val enumDef: EnumDef
+) : StubElementBuilder {
+    override fun build(): List<StubIrElement> {
+        if (!context.isStrictEnum(enumDef)) {
+            return generateEnumAsConstants(enumDef)
+        }
+
+        val baseTypeMirror = context.mirror(enumDef.baseType)
+        val baseType = WrapperStubType(baseTypeMirror.argType)
+
+        val clazz = (context.mirror(EnumType(enumDef)) as TypeMirror.ByValue).valueType.classifier
+
+        val enumVariants = enumDef.constants.map {
+            val literal = context.tryCreateIntegralStub(enumDef.baseType, it.value)
+                    ?: error("Cannot create enum value ${it.value} of type ${enumDef.baseType}")
+            EnumEntryStub(it.name.asSimpleName(), literal)
```

Comment:
All `.asSimpleName()` should be moved to text emitter to avoid
1.  applying it twice
2.  applying it when emitting metadata

Reply:
To ensure that 1 is not the case, consider testing with `asSimpleName()` adding back ticks unconditionally.